### PR TITLE
python: fix venv script permissions

### PR DESCRIPTION
**Pull request by waldyrious**
Preserve write permissions in the [venv activation scripts](https://docs.python.org/3/library/venv.html#how-venvs-work) in the Python formulas (`python@3.14`, `python@3.13`, `python@3.12`, `python@3.11` and `python-freethreading`). This prevents errors when reinitializing an existing venv (e.g. running `python3 -m venv myenv` twice), as described in the upstream CPython issue: https://github.com/python/cpython/issues/86079.

In practice, the permissions of these template scripts are not actually changed in this PR; they are just excluded from automatic scrubbing of their writable bit, by adding a `skip_clean` annotation for their directory, which prevents Homebrew's [Cleaner::clean_dir](https://github.com/Homebrew/brew/blob/966384bd63031be492c30f4a2fa9c100101c65dc/Library/Homebrew/cleaner.rb#L132) method from setting them to be read-only — which is done to all non-executable files.

(These venv activation scripts are not executable because they're meant to be sourced, not executed, so they are treated as plain data files, lacking a shebang and the executable bit.)

See the first commit message (for `python@3.14`) for additional context and details.

-----

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`?
- [x] Is your test running fine `brew test <formula>`?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)?

-----

- [x] AI was used to generate or assist with generating this PR. *Please specify below how you used AI to help you, and what steps you have taken to manually verify the changes*.

  AI (Claude Opus 4.6 via Kiro, GPT 5.4 via Copilot CLI) was used to research context, discuss the approach, and implement the first version of this change (which was based on `chmod`, similar to what was done in [PR #256454](https://github.com/Homebrew/homebrew-core/pull/256454)); later, I also used Opus 4.6 to troubleshoot a CI failure related to that first approach, and identify and implement the `skip_clean` approach instead.

### CODE REVIEW

**Comment by waldyrious** [comment]
Could a maintainer please re-run the `macOS 15-x86_64` CI job? It was canceled in the [last run](https://github.com/Homebrew/homebrew-core/actions/runs/25964837349/job/76326647678?pr=279955), but had passed in a [previous run](https://github.com/Homebrew/homebrew-core/actions/runs/25105432328/job/73566003327?pr=279955).

**Comment by waldyrious** [comment]
Alright, I'll try rebasing and seeing if CI will be able to pass now.

**Comment by waldyrious** [comment]
The [CI failure](https://github.com/Homebrew/homebrew-core/actions/runs/27011878376/job/79732878845) is unrelated: the test for the `streamlink` formula is broken since it currently [relies](https://github.com/Homebrew/homebrew-core/blob/53ff04fe06e73b130c6e38929d996605f2ca18c6/Formula/s/streamlink.rb#L126) on fetching a specific video from Vimeo (https://player.vimeo.com/video/941078932), which has been removed. The fix needs to be done in the streamlink formula, after which this PR should be able to pass CI.

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good! Let's remove the `revision` bumps, though; this doesn't warrant forcing a reinstall for all users.

**Comment by waldyrious** [comment]
Thanks for the review! I've removed the revision bumps and rebased the branch.

File: Formula/p/python-freethreading.rb [hunk 1]
```diff
@@ -418,7 +423,8 @@ def caveats
     # and it can occur that building sqlite silently fails if OSX's sqlite is used.
     system python3, "-c", "import sqlite3"
 
-    # check to see if we can create a venv
+    # Check to see if we can create a venv, and reinitialize it in place.
+    system python3, "-m", "venv", testpath/"myvenv"
```
Comment (MikeMcQuaid) [inline Formula/p/python-freethreading.rb]: This is now duplicated

File: Formula/p/python@3.11.rb [hunk 2]
```diff
@@ -497,7 +502,8 @@ def caveats
     # and it can occur that building sqlite silently fails if OSX's sqlite is used.
     system python3, "-c", "import sqlite3"
 
-    # check to see if we can create a venv
+    # Check to see if we can create a venv, and reinitialize it in place.
+    system python3, "-m", "venv", testpath/"myvenv"
```
Comment (MikeMcQuaid) [inline Formula/p/python@3.11.rb]: and here

File: Formula/p/python@3.12.rb [hunk 3]
```diff
@@ -478,7 +483,8 @@ def caveats
     # and it can occur that building sqlite silently fails if OSX's sqlite is used.
     system python3, "-c", "import sqlite3"
 
-    # check to see if we can create a venv
+    # Check to see if we can create a venv, and reinitialize it in place.
+    system python3, "-m", "venv", testpath/"myvenv"
```
Comment (MikeMcQuaid) [inline Formula/p/python@3.12.rb]: and here

File: Formula/p/python@3.13.rb [hunk 4]
```diff
@@ -464,7 +469,8 @@ def caveats
     # and it can occur that building sqlite silently fails if OSX's sqlite is used.
     system python3, "-c", "import sqlite3"
 
-    # check to see if we can create a venv
+    # Check to see if we can create a venv, and reinitialize it in place.
+    system python3, "-m", "venv", testpath/"myvenv"
```
Comment (MikeMcQuaid) [inline Formula/p/python@3.13.rb]: and here

File: Formula/p/python@3.14.rb [hunk 5]
```diff
@@ -478,7 +483,8 @@ def caveats
     # and it can occur that building sqlite silently fails if OSX's sqlite is used.
     system python3, "-c", "import sqlite3"
 
-    # check to see if we can create a venv
+    # Check to see if we can create a venv, and reinitialize it in place.
+    system python3, "-m", "venv", testpath/"myvenv"
```
Comment (MikeMcQuaid) [inline Formula/p/python@3.14.rb]: and here

File: Formula/p/python-freethreading.rb (hunk 1 above)
Comment (waldyrious) [reply Formula/p/python-freethreading.rb]: That's deliberate :) the issue this PR fixes is precisely the ability to reinitialize an existing venv without getting a permission error, by ensuring the template files that get copied into the new venv keep their write permissions. The test to guarantee that this works is then precisely to run the same venv initialization command twice.

Do you have any suggestion for how to make this clearer, e.g. by a different phrasing of the comment above these lines?

Comment (woodruffw) [reply Formula/p/python-freethreading.rb]: My 0.02c is that this is probably fine to leave out entirely -- our test stanzas are meant to be smoke tests, not placement for regressions/regression fixes that happen over time.

(This is a hard balance to strike of course.)

Comment (waldyrious) [reply Formula/p/python-freethreading.rb]: I'm happy to remove the tests — indeed it's an edge case.

File: Formula/p/python-freethreading.rb [hunk 6]
```diff
@@ -44,6 +44,11 @@ class PythonFreethreading < Formula
     depends_on "zlib-ng-compat"
   end
 
+  # Allow venv activation scripts to keep their writable flag
+  # so that virtual environments can be reinitialized.
+  # See: https://github.com/python/cpython/issues/86079
+  skip_clean "lib/python#{version.major_minor}t/venv/scripts"
```
Comment (woodruffw) [inline Formula/p/python-freethreading.rb]: `skip_clean` does a lot of other things too. Most (all?) of those probably aren't relevant in this directory, but it might be better to do something more targeted here.

Maybe we can explicitly re-add the user write bit to just the activation shims?

Comment (waldyrious) [reply Formula/p/python-freethreading.rb]: > Maybe we can explicitly re-add the user write bit to just the activation shims?

I did try precisely that —i.e. `chmod u+w venv/scripts/**/*`— in the first approach of this PR (see commit [05bff77](https://github.com/Homebrew/homebrew-core/commit/05bff77333c387a78aab2cdc31c57d0d6d712d43)) but that [failed in CI](https://github.com/Homebrew/homebrew-core/actions/runs/25083140207/job/73493072694?pr=279955) with the same permission error as the current formulae. The reason seems to be that the `Cleaner` runs after the `install` method, and at that point it [resets all non-executable files to 0444](https://github.com/Homebrew/brew/blob/966384bd63031be492c30f4a2fa9c100101c65dc/Library/Homebrew/cleaner.rb#L145-L150). I suppose I could try that same approach in the `post_install` step, but considering that only the `/venv/scripts` directory is being exempted from the Cleaner flow, does it make sense to let the cleaner scrub the permissions only to re-add them later? It feels more appropriate to address the issue by preventing the actions that cause it, than by undoing them afterwards.

File: Formula/p/python@3.11.rb (hunk 2 above)
Comment (waldyrious) [reply Formula/p/python@3.11.rb]: Addressed [above](https://github.com/Homebrew/homebrew-core/pull/279955#discussion_r3474160513). Resolving this thread for convenience.

File: Formula/p/python@3.12.rb (hunk 3 above)
Comment (waldyrious) [reply Formula/p/python@3.12.rb]: Addressed [above](https://github.com/Homebrew/homebrew-core/pull/279955#discussion_r3474160513). Resolving this thread for convenience.

File: Formula/p/python@3.13.rb (hunk 4 above)
Comment (waldyrious) [reply Formula/p/python@3.13.rb]: Addressed [above](https://github.com/Homebrew/homebrew-core/pull/279955#discussion_r3474160513). Resolving this thread for convenience.

File: Formula/p/python@3.14.rb (hunk 5 above)
Comment (waldyrious) [reply Formula/p/python@3.14.rb]: Addressed [above](https://github.com/Homebrew/homebrew-core/pull/279955#discussion_r3474160513). Resolving this thread for convenience.

**Comment by waldyrious** [comment]
This is not stale — I'm just waiting on a response by @woodruffw [here](https://github.com/Homebrew/homebrew-core/pull/279955#discussion_r3474901095).

File: Formula/p/python-freethreading.rb (hunk 6 above)
Comment (MikeMcQuaid) [reply Formula/p/python-freethreading.rb]: @waldyrious I agree with @woodruffw here. Can you try this without `skip_clean` and use `post_install_steps` for now.

Comment (waldyrious) [reply Formula/p/python-freethreading.rb]: Alright. Updated the approach. Tested locally, working as expected :)

**Review by MikeMcQuaid — APPROVED** [review]
Thanks!

**Comment by waldyrious** [comment]
@MikeMcQuaid should I rebase the branch as a way to give the CI another chance to run to completion? Or are such failures bound to happen given such a lengthy list of dependents, and therefore not worth the extended wait and use of CI resources — especially for such a (IMHO) minor change?

**Comment by MikeMcQuaid** [comment]
yes, let's skip dependents. rerunning. don't rebase please ❤️
